### ISSUE TITLE
QSelect input now fails WCAG 2.0 A standards

### ISSUE THREAD

**Issue by wreality** [quasarframework/quasar#10127]
**Describe the bug**
The changes introduced by #9983 cause QSelects to actively fail WCAG 2.0 A checks in axe-core now.  Admittedly, the change is more accessible, but now fails standards checks whereas before it was less accessible but passed the check. :-)  

The specific violations:
 - Ensures ARIA attributes are allowed for an element's role: ARIA attribute is not allowed: aria-multiselectable="false"
   - Likely Remedy: `aria-multiselectable` belongs on the element with `role="listbox"`
 - aria-labelledby or aria-label attribute does not exist, references elements that do not exist or references elements that are empty
   - Likely Remedy: Should add `aria-label` similar to how QInput behaves.
 - Invalid ARIA attribute value: aria-activedescendant="f_a1004b26-2180-4fad-8186-a3b48c387fe1_-1"
   - Likely Remedy: `aria-activedescendant` must be a descendant of the element **or in the `aria-owns` comma-separated list**.  The current implementation puts the element with `role="listbox"` in the `aria-owns` attr rather than the individual elements that have `role="option"`.  Use comma-separated list of `role="option"` items in the `aria-owns` attribute. (This one is silly, the implementation here is cleaner, but [this is what the spec](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role) says)

**Codepen/jsFiddle/Codesandbox (required)**
Codepen doesn't really describe the problem.  The described violations occur when running [Axe devtools](https://www.deque.com/axe/) on [quasar.dev](https://v1.quasar.dev/vue-components/select#overview). 

**Expected behavior**
No accessibility violations.

**Platform (please complete the following information):**
Quasar Version: 1.15.23

**Comment by pdanpdan** [comment]
`aria-multiselectable`: agree
`aria-label`: agree, forgot about it

`aria-activedescendant`: I can't make heads or tails of it because of these examples:
- https://www.digitala11y.com/aria-activedescendant-properties/
- https://www.digitala11y.com/aria-owns-properties/
- https://www.digitala11y.com/aria-expanded-state/
- https://www.digitala11y.com/aria-autocomplete-properties/

Also the examples do not seem to agree on the position of the `aria-expanded` attr.
But I'm not going to add a list of all option ids to `aria-owns` no matter what that checker sais :)

I'll make a test build this evening and ask you to test.

**Comment by pdanpdan** [comment]
The changes should be from:

```javascript
    comboboxAttrs () {
      return {
        role: 'combobox',
        'aria-multiselectable': this.multiple === true ? 'true' : 'false',
        'aria-expanded': this.menu === true ? 'true' : 'false',
        'aria-owns': `${this.targetUid}_lb`,
        'aria-activedescendant': `${this.targetUid}_${this.optionIndex}`
      }
    },

    listboxAttrs () {
      return {
        role: 'listbox',
        id: `${this.targetUid}_lb`
      }
    }
```

to:

```javascript
    comboboxAttrs () {
      return {
        role: 'combobox',
        'aria-label': this.label,
        'aria-autocomplete': this.useInput === true ? 'list' : 'none',
        'aria-expanded': this.menu === true ? 'true' : 'false',
        'aria-owns': `${this.targetUid}_lb`,
        'aria-controls': `${this.targetUid}_lb`,
        'aria-activedescendant': `${this.targetUid}_${this.optionIndex}`
      }
    },

    listboxAttrs () {
      return {
        role: 'listbox',
        id: `${this.targetUid}_lb`,
        'aria-multiselectable': this.multiple === true ? 'true' : 'false',
        'aria-activedescendant': `${this.targetUid}_${this.optionIndex}`
      }
    }
```

If you have other suggestions please tell me.
`aria-activedescendant` must be present on the combobox/input or else it won't be picked by the screen reader

**Comment by gmeben** [comment]
@pdanpdan I am willing to assist with testing your build.

**Comment by wreality** [comment]
> But I'm not going to add a list of all option ids to `aria-owns` no matter what that checker sais :)
> 

I agree it seems particularly ugly. I'll take the examples and create an issue on axe-core.

**Comment by wreality** [comment]
The w3 examples for activedescedent and combobox show the attribute on the listbox element. 

https://www.w3.org/TR/wai-aria-practices-1.1/examples/combobox/aria1.1pattern/listbox-combo.html


This may be something to be submitted to the screen reader as a bug if it's not following standards.

**Comment by pdanpdan** [comment]
https://github.com/pdanpdan/quasar/releases/tag/quasar-pdan-v1.15.23-beta.2

In package.json replace quasar dependency with:
"quasar": "https://github.com/pdanpdan/quasar#quasar-pdan-v1.15.23-beta.2"

Then do a yarn or npm install and start the app again.

Please report if you find something not working.

Check:
- with the a11y checker
- with a screen reader (I am on linux, more checks are better)

**Comment by gmeben** [comment]
While testing a page with a QSelect component, my colleague @wreality and I have found a `role="combobox"` element is given an invalid `aria-activedescendant` attribute value (i.e. `#f_5331e417-d2a9-403d-8071-fa228b4fcee9_-1`) when its listbox is not visible. 

<img width="1435" alt="screenshot of axe DevTools featuring the invalid ARIA attribute value violation" src="https://user-images.githubusercontent.com/2668987/126526086-5eb6a1b1-2e2b-46ac-81b6-aa870dfdc2d9.png">

Note the negative one (-1) appended to the value of this attribute.

We believe this attribute should be empty by default and not refer to an element that doesn't exist.

**Comment by pdanpdan** [comment]
Try 
"quasar": "https://github.com/pdanpdan/quasar#quasar-pdan-v1.15.23-beta.3"

**Comment by gmeben** [comment]
@pdanpdan That appears to have fixed it. Thank you so much for your timely contributions.

**Comment by pdanpdan** [comment]
I need to know what you have tested (validator / screen reader), what os and what browser.
I do my tests in chrome + firefox on linux, with the default screen reader.
If all is well I'll make the PR.

**Comment by gmeben** [comment]
Good point. 

I've been using axe DevTools (which uses axe-core 4.3.1) in Chrome 91 on macOS 11.4 to do the a11y validation. 

I will get back to you on screen reader testing.

**Comment by gmeben** [comment]
I did some screen reader testing with the [Screen Reader extension](https://chrome.google.com/webstore/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn/related) (formerly known as ChromeVox) version 53.0.2784.12 for Chrome 92 on macOS 11.4.

### Keyboard

1. When I press the `tab` key to focus a QSelect with no option selected, Screen Reader reads (in order): 
  - "Select Fruit" - which is the `label` attribute 
  - "combobox collapsed" 
  - "main" - which I presume is the context of the QSelect within the page 
2. When I press the `enter` key, the listbox options become visible and Screen Reader reads: 
  - "expanded" 
3.  I'm able to use the `up`/`down` arrow keys to move focus between options. When I do so, Screen Reader reads: 
  - "Apple" - which is the currently focused option text 
  - "Not selected" 
  - "1 of 6" - which is the currently focused option's placement within the total count of options 
4. I can press the `enter` key to select an option. Screen Reader reads: 
  - "Select Fruit" - which is the `label` attribute 
  - "combobox collapsed" 
  - "main" - which I presume is the context of the QSelect within the page 
5. When I `tab` away to focus on another element and `tab` back to re-focus the QSelect input, Screen Reader reads: 
  - "Select Fruit" - which is the `label` attribute 
  - "combobox collapsed" 
  - "main" - which I presume is the context of the QSelect within the page 
 6. When I press the `enter` key, Screen Reader reads: 
  - "Apple" - which is the currently selected option text 
  - "Selected" 
  - "1 of 6" - which is the currently selected option's placement within the total count of options 

### Mouse

7. When I click to focus a QSelect with no option selected, Screen Reader reads (in order): 
  - "Select Fruit" - which is the `label` attribute 
  - "combobox expanded" 
8. When I click an option to select it, Screen Reader reads: 
  - "Select Fruit" - which is the `label` attribute 
  - "combobox collapsed" 
  - "main" - which I presume is the context of the QSelect within the page 
9. When I click elsewhere to focus on another element and click back to re-focus the QSelect input, Screen Reader reads: 
  - "Select Fruit" - which is the `label` attribute 
  - "Apple" - which is the currently selected option text 
  - "combobox selected" 

### Focus Indicator

<img width="1011" alt="screenshot of Chome DevTools opened on a sample web application where Screen Reader is focused on a QSelect component" src="https://user-images.githubusercontent.com/2668987/126677183-99f46b67-3a7f-4881-aa3b-9d066efa073e.png">

Screen Reader behaves strangely with its orange focus indicator on QSelect components. When the input of a QSelect is focused, Screen Reader positions its indicator at the end of the selected option's text and, rather than fully wrapping around the text, the indicator appears like a small rounded square floating on top of the text. 

I'm not sure to what degree Screen Reader or Quasar is responsible for this unfavorable behavior but I'm willing to concede that it's not entirely Quasar. 

### Conclusion

I find the Screen Reader accessibility and user experience of the QSelect component acceptable. However, item 8 (clicking on an option with a mouse to select it in a QSelect component) leaves something to be desired.

**Comment by pdanpdan** [comment]
Wow, very nice report. Let me digest it over the weekend and I'll be back with news. Thank you.

**Comment by pdanpdan** [comment]
Hello.
I've checked point 8 with the system screen reader from linux and it looks like it reads the clicked option fine - so maybe it's a quirk with the screen reader plugin for chrome.
And about the focus indicator - I know what it is, but I can think of no method to fix it - the QSelect that does not use input has a placeholder input to receive the focus for autocomplete, and it is empty and of 0 width - and that is what is highlighted
But I'm afraid that any method I use to hide it will hinder the screen reader - so for the moment I think this is what I can do.

Thank you all for the help and testing, it was a nice experience.

**Comment by pdanpdan** [comment]
Btw, if someone wants we would also need some help with https://github.com/quasarframework/quasar/issues/9689
But in that case it's a little more complex because I can see no way to avoid adding localized strings.
I'll add more details there.

**Comment by rstoenescu** [comment]
Fix will be available in Quasar v2.1.6.

**Comment by gaby64** [comment]
this appears to break keyboard navigation through forms, I cant tab to my qselects

**Comment by pdanpdan** [comment]
Hello. I was working on some changes for a11y and I need some help with testing.

The changes are related to:
- feat(QExpansionItem): add a11y for toggle buttons; improve header slot scope quasarframework#14338
- feat(QSelect): improve a11y - aria-readonly on focusable element; remove aria-activedescendant duplicate
- feat(QChip): add aria-label and proper role for remove button/icon
- feat(QDialog): set aria-modal on dialog
- feat(QMenu): set default role of menu
- feat(QFab): add a11y for toggle button; describe aria-controls/own
- feat(QBtnDropdown): add a11y for toggle buttons; describe aria-controls/own

I have published a test documentation site at https://pdanpdan.github.io/quasar-docs/ using the changed code.
If you use screen readers and/or a11y checking tools can you please check the pages for the changed components and report if there are problems or if there are improvements?
From the docs you can open codepens with the changed code, so if you have special cases you want to test you can.
